### PR TITLE
Hw31

### PR DESCRIPTION
**Pull request by frostpv**

### CODE REVIEW

File: src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+            preparedStatement.setString(1, item.getName());
+            preparedStatement.setDouble(2, item.getPrice());
+            preparedStatement.executeUpdate();
+            ResultSet resultSet = preparedStatement.getGeneratedKeys();
+            if (resultSet.next()) {
+                item.setId(resultSet.getLong(1));
+            }
+            return item;
+        } catch (SQLException e) {
+            throw new RuntimeException("Product " + item + " was not created", e);
```
Comment (Alkey) [inline src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java]: Let’s create custom exception DataProcessingException with two parameters: String message and Throwable ex.

File: src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+            ResultSet resultSet = preparedStatement.executeQuery();
+            if (resultSet.next()) {
+                long productId = resultSet.getLong("product_id");
+                String name = resultSet.getString("name");
+                double price = resultSet.getDouble("price");
+                Product product = new Product(name, price);
+                product.setId(productId);
+                return Optional.of(product);
+            }
+            throw new RuntimeException(exceptionMessage);
```
Comment (Alkey) [inline src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java]: same

File: src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+            String query = "UPDATE products SET name = ?, price = ? WHERE product_id = ?";
+            PreparedStatement preparedStatement
+                    = connection.prepareStatement(query, Statement.RETURN_GENERATED_KEYS);
+            preparedStatement.setString(1, item.getName());
+            preparedStatement.setDouble(2, item.getPrice());
+            preparedStatement.setLong(3, item.getId());
+            preparedStatement.executeUpdate();
+            return item;
+        } catch (SQLException e) {
+            throw new RuntimeException("Product " + item + " not updated", e);
```
Comment (Alkey) [inline src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java]: same

File: src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+    @Override
+    public boolean delete(Long item) {
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            String query = "UPDATE products SET deleted = 1 WHERE product_id = ?";
+            PreparedStatement preparedStatement = connection.prepareStatement(query);
+            preparedStatement.setLong(1, item);
+            return preparedStatement.executeUpdate() == 1;
+        } catch (SQLException e) {
+            throw new RuntimeException("Product with id - " + item + " not deleted");
```
Comment (Alkey) [inline src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java]: same

File: src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+            while (resultSet.next()) {
+                Long productId = resultSet.getLong("product_id");
+                String name = resultSet.getString("name");
+                double price = resultSet.getDouble("price");
+                Product product = new Product(name, price);
+                product.setId(productId);
+                products.add(product);
+            }
+        } catch (SQLException e) {
+            throw new RuntimeException("All products error", e);
```
Comment (Alkey) [inline src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java]: same

File: src/main/java/com/internet/shop/WebShopApp.java [hunk 6]
```diff
@@ -1,65 +1,16 @@
...
 
 public class WebShopApp {
     public static void main(String[] args) {
-        ProductService productService =
-                (ProductService) Injector.getInstance("com.internet.shop")
-                 .getInstance(ProductService.class);
+        ProductDaoJdbcImpl daoJdbc = new ProductDaoJdbcImpl();
+        Product product = new Product("BMW M5", 54555);
+        System.out.println(daoJdbc.create(product).toString());
+        System.out.println(daoJdbc.get(1L).toString());
```
Comment (Alkey) [inline src/main/java/com/internet/shop/WebShopApp.java]:
```suggestion
        System.out.println(daoJdbc.get(1L).toString());
        System.out.println(daoJdbc.get(product.getId).toString());
```

File: src/main/java/com/internet/shop/WebShopApp.java [hunk 7]
```diff
@@ -1,65 +1,16 @@
...
+        daoJdbc.delete(1L);
```
Comment (Alkey) [inline src/main/java/com/internet/shop/WebShopApp.java]:
```suggestion
        daoJdbc.delete(1L);
        daoJdbc.delete(product.getId);
```

File: src/main/java/com/internet/shop/WebShopApp.java [hunk 8]
```diff
@@ -1,65 +1,16 @@
...
+        System.out.println(daoJdbc.getAll());
```
Comment (Alkey) [inline src/main/java/com/internet/shop/WebShopApp.java]: What about update method?

File: src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java (hunk 1 above)
Comment (frostpv) [reply src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java]: added

File: src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java (hunk 2 above)
Comment (frostpv) [reply src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java]: fixed

File: src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java (hunk 3 above)
Comment (frostpv) [reply src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java]: fixed

File: src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java (hunk 4 above)
Comment (frostpv) [reply src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java]: fixed

File: src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java (hunk 5 above)
Comment (frostpv) [reply src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java]: fixed

File: pom.xml [hunk 9]
```diff
@@ -37,6 +37,12 @@
             <version>4.12</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>mysql</groupId>
+            <artifactId>mysql-connector-java</artifactId>
+            <version>8.0.17</version>
+        </dependency>
+
```
Comment (Sofasmile) [inline pom.xml]: let's remove empty line

File: src/main/java/com/internet/shop/WebShopApp.java [hunk 10]
```diff
@@ -1,65 +1,20 @@
...
-        shoppingCartService.addProduct(jonCart, iphone6s);
-        shoppingCartService.addProduct(jonCart, iphone7plus);
-        System.out.println(shoppingCartService.getByUserId(jon.getId()));
-        OrderService orderService =
-                (OrderService) Injector.getInstance("com.internet.shop")
-                        .getInstance(OrderService.class);
-        orderService.completeOrder(jonCart);
-        System.out.println(shoppingCartService.getByUserId(jon.getId()));
-        System.out.println(orderService.getAll());
+        ProductDaoJdbcImpl daoJdbc = new ProductDaoJdbcImpl();
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/WebShopApp.java]: can we use injector here?

File: src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+            PreparedStatement preparedStatement = connection.prepareStatement(query);
+            preparedStatement.setLong(1, item);
+            ResultSet resultSet = preparedStatement.executeQuery();
+            if (resultSet.next()) {
+                long productId = resultSet.getLong("product_id");
+                String name = resultSet.getString("name");
+                double price = resultSet.getDouble("price");
+                Product product = new Product(name, price);
+                product.setId(productId);
+                return Optional.of(product);
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java]: https://mate-academy.github.io/jv-program-fulltime/02_jdbc_and_servlets/checklist/hw_21.html
first point

File: src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+            }
+            return Optional.empty();
+        } catch (SQLException e) {
+            throw new DataBaseProcessingException("Can`t get product with id", e);
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java]: let's add id

File: src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+        }
+    }
+
+    @Override
+    public Product update(Product item) {
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            String query = "UPDATE products SET name = ?, price = ? WHERE product_id = ?";
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java]: can we update deleted data?

File: src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+@Dao
+public class ProductDaoJdbcImpl implements ProductDao {
+    @Override
+    public Product create(Product item) {
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java]:
```suggestion
    public Product create(Product product) {
```

File: src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+                item.setId(resultSet.getLong(1));
+            }
+            return item;
+        } catch (SQLException e) {
+            throw new DataBaseProcessingException("Product " + item + " was not created", e);
+        }
+    }
+
+    @Override
+    public Optional<Product> get(Long item) {
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java]:
```suggestion
    public Optional<Product> get(Long id) {
```

File: pom.xml (hunk 9 above)
Comment (frostpv) [reply pom.xml]: fixed

File: src/main/java/com/internet/shop/WebShopApp.java (hunk 10 above)
Comment (frostpv) [reply src/main/java/com/internet/shop/WebShopApp.java]: used

File: src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java (hunk 11 above)
Comment (frostpv) [reply src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java]: fixed

File: src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java (hunk 12 above)
Comment (frostpv) [reply src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java]: fixed

File: src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java (hunk 13 above)
Comment (frostpv) [reply src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java]: fixed

File: src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java (hunk 14 above)
Comment (frostpv) [reply src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java]: fixed

File: src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java (hunk 15 above)
Comment (frostpv) [reply src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java]: fixed

File: src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java [hunk 16]
```diff
@@ -91,17 +88,20 @@ public boolean delete(Long item) {
...
+                products.add(fetchProduct(resultSet).get());
             }
         } catch (SQLException e) {
             throw new DataBaseProcessingException("All products error", e);
         }
         return products;
     }
-}
 
+    private Optional<Product> fetchProduct(ResultSet resultSet) throws SQLException {
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java]:
```suggestion
    private Product fetchProduct(ResultSet resultSet) throws SQLException {
```

Comment (frostpv) [reply src/main/java/com/internet/shop/dao/ProductDaoJdbcImpl.java]: fixed

**Review by Sofasmile — APPROVED** [review]

**Pull request merged**